File: Portfolio/Pages/Home.razor.ts
// Interface für das disposable Objekt, das vom ScrollSpy zurückgegeben wird.
interface Disposable {
    dispose(): void;
}

// Interface für das BoundingClientRect Objekt.
interface BoundingClientRect {
    height: number;
}

/**
 * Temporäre Debugging-Version 3: Initialisiert den Scroll-Spy OHNE Event-Listener.
 * RUFT DEN HANDLER NUR EINMAL INITIAL AUF. Loggt Parameter beim Eintritt.
 * @param sectionIds - Erwartet Array von IDs.
 * @param navbarHeight - Erwartet Zahl.
 * @param dotNetHelper - Erwartet Zahl.
 * @returns Ein Dummy-Disposable-Objekt oder null.
 */
// Behalten Sie die Signatur der Test Version 2 bei (2 Parameter)
export function initScrollSpy(dotNetHelper: any, sectionIds: string[], navbarHeight: number): Disposable | null {
    
    // <-- HIER TRITT DER FEHLER 'sectionIds.map is not a function' AUF, WENN sectionIds KEIN ARRAY IST -->
    const sections: HTMLElement[] = sectionIds
        .map(id => document.getElementById(id))
        .filter((el): el is HTMLElement => el != null);

    if (sections.length === 0) {
        console.warn("ScrollSpy: Keine Sektionen für die gegebenen IDs gefunden."); // Behalte den Warn-Log
        return null;
    }

    // Die scrollHandler Funktion bleibt gleich (ohne dotNetHelper Aufruf)
    const scrollHandler = () => {
        let currentSectionId: string | null = null;
        const scrollY: number = window.scrollY;
        
        // ... Der Rest der scrollHandler Logik (Berechnung der aktiven Sektion, ohne dotNetHelper Aufruf) bleibt gleich ...
        for (let i = sections.length - 1; i >= 0; i--) {
            const section = sections[i];
            if (section && section.offsetHeight > 0) {
                // Verwenden Sie hier die korrekten Parameter aus der Funktionssignatur
                const sectionTop: number = section.offsetTop - navbarHeight - 1;

                if (scrollY >= sectionTop) {
                    currentSectionId = section.getAttribute('id');
                    break;
                }
            }
        }

        if (!currentSectionId && sections.length > 0) {
            currentSectionId = sections[0].getAttribute('id');
        }

        if (currentSectionId !== null) {
            // Überprüfen Sie, ob dotNetHelper verfügbar ist (sicherheitshalber)
            if (dotNetHelper && dotNetHelper.invokeMethodAsync) {
                // Rufe die C# Methode SetActiveSection auf und übergebe die Sektions-ID
                dotNetHelper.invokeMethodAsync('SetActiveSection', currentSectionId)
                    .catch((error: any) => {
                        console.error("Fehler beim Aufrufen von SetActiveSection:", error);
                    });
            } else {
                console.warn("DotNet Helper ist nicht verfügbar...");
            }
        }
    };

    // Event Listener registrieren
    window.addEventListener("scroll", scrollHandler);
    
    // RUFT DEN HANDLER NUR EINMAL SOFORT AUF (KEIN EVENT-LISTENER REGISTRIERT)
    scrollHandler();
    
    // Gibt ein Dummy Disposable Objekt zurück
    return {
        dispose: () => {
            window.removeEventListener("scroll", scrollHandler);
        }
    };
}

/**
 * Scrollt das Fenster sanft zu einer Zielsektion. DIESE FUNKTION WIRD EXPORTIERT.
 * @param sectionId - Die ID der Sektion.
 * @param navbarHeight - Die Höhe der Navbar.
 */
export function scrollToSection(sectionId: string, navbarHeight: number): void {
    const targetElement = document.getElementById(sectionId);
    if (targetElement) {
        window.scrollTo({
            top: targetElement.offsetTop - navbarHeight,
            behavior: 'smooth'
        });
    }
}

/**
 * Holt das Bounding Client Rectangle eines Elements. DIESE FUNKTION WIRD EXPORTIERT.
 * @param element - Das HTML-Element.
 * @returns Das BoundingClientRect oder null.
 */
export function getElementBoundingClientRect(element: HTMLElement | null | undefined): BoundingClientRect | null {
    if (element) {
        const rect = element.getBoundingClientRect();
        return {
            height: rect.height
        };
    }
    return null;
}

File: Portfolio/Pages/Home.razor.cs
using System.Text.Json.Serialization;
using Microsoft.AspNetCore.Components;
using Microsoft.JSInterop;
using Portfolio.Models;
using Portfolio.Services;

namespace Portfolio.Pages;

public partial class Home : ComponentBase, IDisposable
{
    [Inject]
    public IJSRuntime JsRuntime { get; set; } = null!;

    [Inject]
    public GitHubService GitHubService { get; set; } = null!;

    [Inject]
    public ILogger<Home> Logger { get; set; } = null!;


    // --- State for Projects Section ---
    private List<GitHubProject> _projects = [];
    private bool _isLoadingProjects = true;
    private string? _projectsErrorMessage;

    // --- State for Navbar ---
    private bool _isNavbarCollapsed = true;
    private string _activeNavLink = "home";
    private ElementReference _navbarCollapseElement;
    private ElementReference _navbarElement;


    // --- State for JS Interop ---
    // Referenz zum geladenen JavaScript-Modul (für Isolation)
    private IJSObjectReference? _module;
    // Referenz zum ScrollSpy-Handler innerhalb des Moduls
    private IJSObjectReference? _scrollSpyHandler;
    // DotNetObjectReference zum Aufrufen von C# aus JS
    private DotNetObjectReference<Home>? _dotNetHelper;
    private static readonly string[] Args = ["home", "about", "projects", "skills", "contact"];


    // --- Lifecycle Methods ---

    protected override async Task OnInitializedAsync()
    {
        await FetchGitHubProjects("Riksorax");
    }

    protected override async Task OnAfterRenderAsync(bool firstRender)
    {
        if (firstRender)
        {
            _dotNetHelper = DotNetObjectReference.Create(this);

            // --- Laden des isolierten JavaScript-Moduls ---
            // der Pfad ist ./_content/[YourAssemblyName]/[KomponentenPfad].razor.js
            // YourAssemblyName ist der Name Ihrer.csproj Datei (z.B. "Portfolio")
            // KomponentenPfad ist der Pfad ab dem Wurzelverzeichnis der Komponenten (z.B. "Pages/Home")
            const string modulePath = "./Pages/Home.razor.js";

            try
            {
                Logger.LogInformation($"Loading JS module from: {modulePath}"); // Optional
                _module = await JsRuntime.InvokeAsync<IJSObjectReference>("import", modulePath);
                Logger.LogInformation("JS module loaded successfully."); // Optional
            }
            catch (Exception ex)
            {
                Logger.LogError(ex, $"Fehler beim Laden des JS-Moduls von {modulePath}"); // Optional
                // Hier könnten Sie auch einen Fehler auf der Seite anzeigen, da JS Interop fehlschlägt
                 _projectsErrorMessage = "Konnte erforderliche JavaScript-Funktionen nicht laden."; // Beispiel
                 _isLoadingProjects = false; // Ladezustand beenden
                 StateHasChanged(); // UI aktualisieren
            }


            // --- Wenn Modul erfolgreich geladen wurde, seine exportierten Funktionen aufrufen ---
            if (_module != null)
            {
                 // Warten Sie, bis das Navbar-Element gerendert ist und erhalten Sie dessen Höhe
                var navbarHeight = 0;
                // Greifen Sie auf das DOM-Element über die ElementReference zu
                try
                {
                    // Rufen Sie die Funktion getElementBoundingClientRect INNERHALB des geladenen Moduls auf
                    // Verwenden Sie module.InvokeAsync, NICHT JsRuntime.InvokeAsync global
                    var rect = await _module.InvokeAsync<BoundingClientRect>("getElementBoundingClientRect", _navbarElement);
                    navbarHeight = (int)rect.Height;
                    Logger.LogInformation($"Navbar height obtained: {navbarHeight}"); // Optional
                }
                catch (Exception ex)
                {
                    Logger.LogError(ex, "Fehler beim Abrufen der Navbar-Höhe aus dem Modul."); // Optional
                    // Fallback oder Fehlerbehandlung, falls der Aufruf fehlschlägt
                }


                // Initialisieren Sie den Scroll-Spy INHALB des Moduls
                // Rufen Sie initScrollSpy INNERHALB des geladenen Moduls auf
                // Übergeben Sie die .NET Referenz und andere Daten
                // Verwenden Sie module. InvokeAsync, NICHT JsRuntime. InvokeAsync global
                _scrollSpyHandler = await _module.InvokeAsync<IJSObjectReference>("initScrollSpy", _dotNetHelper, Args, navbarHeight); // <-- Stellen Sie sicher, dass dies der Aufruf ist
                Logger.LogInformation("ScrollSpy initialized successfully."); // Optional


                // Aktualisieren Sie das Jahr im Footer über eine JS Interop (kann weiterhin global sein oder ins Modul verschoben werden)
                await JsRuntime.InvokeVoidAsync("eval", $"document.getElementById('year').textContent = new Date().getFullYear();");
            }
            else
            {
                 // Das Modul konnte nicht geladen werden. Behandeln Sie dies.
                 Logger.LogError("JS module is null, cannot initialize features relying on it."); // Optional
                 // Hier könnten Sie UI-Elemente deaktivieren oder eine Fehlermeldung anzeigen.
            }
        }
    }

     // Implementiert IDisposable, um den JS-Event-Listener und die DotNetObjectReference aufzuräumen
     public async void Dispose()
     {
         // Dispose des ScrollSpy-Handlers INNERHALB des Moduls
         // dies muss AUF dem Modul aufgerufen werden, wenn der Handler im Modul liegt
         if (_scrollSpyHandler != null)
         {
             try
             {
                 // Rufen Sie die Dispose-Methode im JS auf dem Handler-Objekt auf
                 await _scrollSpyHandler.InvokeVoidAsync("dispose");
                 // Entsorgen Sie die JS Objekt-Referenz
                 await _scrollSpyHandler.DisposeAsync();
                 Logger.LogInformation("ScrollSpy handler disposed."); // Optional
             }
             catch (JSDisconnectedException)
             {
                 /* Ignorieren, wenn die JS-Runtime bereits getrennt ist */
             }
             catch (Exception ex)
             {
                 Logger.LogError(ex, $"Fehler beim Entsorgen des ScrollSpy-Handlers.");
             }
         }

         // Entsorgen Sie die DotNetObjectReference
         _dotNetHelper?.Dispose();
         Logger.LogInformation("DotNetObjectReference disposed."); // Optional

         // Dispose des JS-Moduls selbst (optional, aber gute Praxis)
         if (_module == null) return;
         {
            try
            {
                await _module.DisposeAsync();
                Logger.LogInformation("JS module disposed."); // Optional
            }
            catch (JSDisconnectedException)
            {
                /* Ignorieren */
            }
            catch (Exception ex)
            {
                Logger.LogError(ex, $"Fehler beim Entsorgen des JS-Moduls.");
            }
        }
    }

    private async Task FetchGitHubProjects(string username)
    {
        _isLoadingProjects = true;
        _projectsErrorMessage = null;
        _projects.Clear();

        try
        {
            _projects = await GitHubService.GetPublicProjectsAsync(username);
        }
        catch (Exception ex)
        {
            _projectsErrorMessage = ex.Message;
            Logger.LogError(ex, "Fehler beim Abrufen der GitHub-Projekte."); // Optional
        }
        finally
        {
            _isLoadingProjects = false;
            StateHasChanged(); // UI aktualisieren
        }
    }

    private void ToggleNavbar()
    {
        _isNavbarCollapsed = !_isNavbarCollapsed;
    }

    // Wird durch @onclick="() => NavigateToSection(...)" in Home.razor ausgelöst
    private async Task NavigateToSection(string sectionId)
    {
        if (!_isNavbarCollapsed)
        {
            _isNavbarCollapsed = true;
        }

        _activeNavLink = sectionId;
        StateHasChanged(); // UI sofort aktualisieren

        // Rufen Sie die scrollToSection Funktion INNERHALB des geladenen Moduls auf
        // Verwenden Sie module. InvokeVoidAsync, NICHT JsRuntime. InvokeVoidAsync global
         if (_module != null) // Prüfen, ob Modul geladen wurde
         {
             var navbarHeight = 0;
             // Erneut die Höhe holen, idealerweise mit der Funktion aus dem Modul
             if (_navbarElement.Id != null)
             {
                 // Rufen Sie die Funktion getElementBoundingClientRect INNERHALB des geladenen Moduls auf
                 var rect = await _module.InvokeAsync<BoundingClientRect>("getElementBoundingClientRect", _navbarElement);
                 navbarHeight = (int)rect.Height;
             } else {
                 // Fallback-Aufruf, falls ElementReference fehlschlägt
                 navbarHeight = await JsRuntime.InvokeAsync<int>("eval", "document.querySelector('.navbar').offsetHeight");
             }

             // Rufen Sie die Funktion scrollToSection INNERHALB des geladenen Moduls auf
             await _module.InvokeVoidAsync("scrollToSection", sectionId, navbarHeight);
         } else {
             Logger.LogError("JS-Modul nicht geladen, kann nicht scrollen."); // Optional
             // Fallback: Standard-Link-Verhalten zulassen oder Fehlermeldung anzeigen
         }
    }

    // Diese Methode wird von JavaScript über [JSInvokable] aufgerufen
    [JSInvokable]
    public void SetActiveSection(string sectionId)
    {
        // Logik zum Aktualisieren des aktiven Links bleibt hier
        _activeNavLink = sectionId;
        StateHasChanged(); // UI aktualisieren
    }

     // --- JS Interop Helper DTO ---
     public class BoundingClientRect
     {
         [JsonPropertyName("height")]
         public double Height { get; init; }
     }
}